include/envoy/server/transport_socket_config.h

```diff
@@ -0,0 +1,66 @@
+#pragma once
+
+#include <string>
+
+#include "envoy/network/transport_socket.h"
+#include "envoy/ssl/context_manager.h"
+
+#include "common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Server {
+namespace Configuration {
+
+/**
+ * Context passed to transport socket factory to access server resources.
+ */
+class TransportSocketFactoryContext {
+public:
+  virtual ~TransportSocketFactoryContext() {}
+
+  /**
+   * @return Ssl::ContextManager& the SSL context manager
+   */
+  virtual Ssl::ContextManager& sslContextManager() PURE;
+
+  /**
+   * @return Stats::Scope& the tansprot socket's stats scope.
+   */
+  virtual Stats::Scope& statsScope() const PURE;
+};
+
+class TransportSocketConfigFactory {
+public:
+  virtual ~TransportSocketConfigFactory() {}
+
+  /**
+   * Create a particular transport socket factory implementation. If the implementation is unable
+   * to produce a factory with the provided parameters, it should throw an EnvoyException. The
+   * returned TransportSocketFactoryPtr should not be nullptr.
```

Comment:
Hm, I see several API docs which have ASCII docs on throwing and several which use "@throw" I mildly prefer the latter since it makes exceptions stand out though there's apparently contention in the javadoc community if @throw is a useful thing.

Reply:
Done